### ISSUE TITLE
Introduce acceptance test framework

### ISSUE THREAD

**Issue by hferentschik** [minishift/minishift#135]
My choice would be [godog](https://github.com/DATA-DOG/godog) which is a Cucumber port for Golang.

Another option could be [Ginko](http://onsi.github.io/ginkgo/), but imo it misses the clarity of Cucumber.

**Comment by jimmidyson** [comment]
Ginkgo is more popular in go world & used in both Kubernetes & OpenShift. I personally much prefer the Go DSL of Ginkgo.

**Comment by hferentschik** [comment]
> I personally much prefer the Go DSL of Ginkgo.

Looks very much like minispec to me and imo a bit too "technical". The benefit of Cucumber imo is that the feature files are written in a human readable fashion which makes the tests understandable even without knowing about Go. Actual coding is separated into step definitions.

**Comment by jimmidyson** [comment]
In reality that very rarely happens though, right? Anyone writing a specification is going to know Go...

**Comment by hferentschik** [comment]
> Anyone writing a specification is going to know Go...

Sure, for the writing part. But maybe not for the reading part. I like the idea that someone can make sense of the tests even w/o knowing Go. Much harder with Ginko. IMO it is much more of a unit test framework opposed to a BDD tool. Cucumber is one abstraction level higher by separating test specification (features) from test code. Obviously this is a matter of taste as well.

**Comment by budhrg** [comment]
Few thoughts before chosing the framework.

I found that `godog` is not mature enough at the moment and even the README says _The API is likely to change a few times before we reach 1.0.0_. Also, found that its documentation is very concise as against to `Ginkgo`.

**Comment by LalatenduMohanty** [comment]
> Ginkgo is more popular in go world & used in both Kubernetes & OpenShift
 +1. 


> I like the idea that someone can make sense of the tests even w/o knowing Go. Much harder with Ginko

But in reality most of the time we have to know go code for a new type of test. So I do not think this is true in for all tests.

**Comment by hferentschik** [comment]
> But in reality most of the time we have to know go code for a new type of test.

no you don't. That's the beauty. Once you have some re-usable steps you can write tests without any Go knowledge. Also it is the amount of noise to signal ratio with Ginko.

We can try both approaches and compare the results.

**Comment by gbraad** [comment]
Currently looking into godog; personally I like the gherkin language to describe your tests, but I do notice that godog lacks the actual given-when-then as part of the internal interface. Everything needs to be setup as a step inside the FeatureContext. This feels less elegant than other implementations.

**Comment by hferentschik** [comment]
> Currently looking into godog; 

ok

> personally I like the gherkin language to describe your tests, 

same here

> but I do notice that godog lacks the actual given-when-then as part of the internal interface. Everything needs to be setup as a step inside the FeatureContext. 

I see. So no true Gherkin DSL.


> This feels less elegant than other implementations.

Maybe you can share your branch / wip pull request at some stage, so that we can get a feel for it as well.

**Comment by gbraad** [comment]
@hferentschik Took the current integration tests as a baseline: https://github.com/gbraad/minishift/tree/bdd-test/test/integration (squashed)
Still haven't fixed running multiple scenarios and some race-condition that occurred.

**Comment by gbraad** [comment]
Combining into a big scenario at the moment would be something like:

```gherkin
  Scenario: Docker environment variables
    Given a "developer"
     When executing "start --vm-driver=virtualbox --docker-env=FOO=BAR --docker-env=BAZ=BAT"
      And should be running
     Then should have a valid IP address
     Then executing "ssh echo hello" should contain
      """
      hello
      """
     Then executing "ssh cat /var/lib/boot2docker/profile" should contain
      """
      export "FOO=BAR"
      export "BAZ=BAT"
      """
```

Current results when running (factor in network conditions):
```
$ go test -timeout 3600s github.com/minishift/minishift/test/integration --tags=integration -v
...... 6


1 scenarios (1 passed)
6 steps (6 passed)
3m24.992346645s
testing: warning: no tests to run
PASS
ok      github.com/minishift/minishift/test/integration 204.996s
$ go test -timeout 3600s github.com/minishift/minishift/test/integration --tags=integration -v
...... 6


1 scenarios (1 passed)
6 steps (6 passed)
2m55.985524022s
testing: warning: no tests to run
PASS
ok      github.com/minishift/minishift/test/integration 175.989s
```

**Comment by hferentschik** [comment]
> Code is not refactored yet, but have made it possible to run multiple scenarios. 

Cool. 

> Also did a quick test if re-using the environment is possible

It should be. It depends what you are after regarding test. Most of the tests we are interested in should be able to re-use the same environment. There might be some tests which are directly affected by the environment, there we need to re-create. 

That said, there is also a possibility to have both. What takes a long time when bootstrapping from nothing are the various downloads. One could have a pre-test task which downloads and caches ISO, oc (and OpenShift image). Then prior each test the artifacts get copied into the right place. I think this would be nice and give us best of both worlds. Unfortunately, we don't quite have the ability yet to cache the Openshift image - see issue #143 and issue #273. 

> The problem lies mostly in doing a reconfigure with docker-env variables. As they are ATM only reset after a minishift delete

Hmm, what kind of variables are we talking about here. Sure we want this to work, but there should nothing be required from our perspective, right?

> Combining into a big scenario at the moment would be something like:

Looks nice. When it comes to readability, I'd expect something like:

    Scenario: Docker environment variables
      Given a "developer"
       When executing "minishift start --vm-driver=virtualbox --docker-env=FOO=BAR --docker-env=BAZ=BAT"
       Then VM should be running
       And should have a valid IP address
       And executing "ssh echo hello" should contain
        """
        hello
        """
       And executing "ssh cat /var/lib/boot2docker/profile" should contain
        """
        export "FOO=BAR"
        export "BAZ=BAT"
        """

Overall, what do you think. Is this the right path to take? Happy with godog?

**Comment by gbraad** [comment]
@hferentschik 

The current reasoning about the syntax was because of
```gherkin
   When executing "minishift start --vm-driver=virtualbox --docker-env=FOO=BAR --docker-env=BAZ=BAT"
```
```gherkin
   And executing "ssh echo hello" should contain
```

the `ssh` is actually used as a subcommand from `minishift`. Although I wanted to simplify, as the actual `minishiftrunner` and `ocrunner` are merely wrappers around `RunCommand` and provide nothing special. We want to assume `minishift`, `oc` and/or `kubectl` to be in the path when executing.

Else,

```gherkin
   When executing minishift command "start ..."
```
and
```gherkin
   When executing openshift command "new-app ..."
```
would be a possible syntax to restrict.  WDYT?

**Comment by gbraad** [comment]
@hferentschik Although, a more generalized `RunCommand` actually makes the syntax and intention clearer.

For:
```gherkin
   Then VM should be running
```
is actually a question that can be answered by the virtual provider, or stick with `minishift` answering this (as currently implemented)?

**Comment by gbraad** [comment]
Cut my PTO short and started to implement a basic test with Ginkgo.  In half an hour I got the following https://github.com/gbraad/minishift/tree/bdd-test2 and some initial points:

1. Very technical way to describe tests
2. Omega (the matcher library) is great as it provides a fluent syntax
3. Although it makes the code prone to brace checking
4. It also feels that code need to be extracted in more util functions that match the return values
5. ex. Valid IP address is not checked against 'nil', but more logical it should be a bool. Non-essential to the actual code under test.

I will add more functionality and refactor later today, but the output is as follows: https://gist.github.com/gbraad/536db4ff31c0559b24926a9d125ed8d2

**Comment by budhrg** [comment]
@gbraad , :+1: 

Does it mean you are evaluating ginko ATM or decided to go with it?

**Comment by gbraad** [comment]
@budhrg evaluating..

**Comment by gbraad** [comment]
Using ginkgo would result into something like:
```
var _ = Describe("Minishift Integration", func() {
    runner: = util.MinishiftRunner {
        Args: * args,
        BinaryPath: * binaryPath
    }

    BeforeSuite(func() {
        testDir: = setUp()
        os.RemoveAll(testDir)
    })

    AfterSuite(func() {
        runner.EnsureDeleted()
    })

    Context("When Minishift is started", func() {
        It("should be started", func() {
            command: = "start --vm-driver=virtualbox --docker-env=FOO=BAR --docker-env=BAZ=BAT"
            runner.RunCommand(command, true)
            Expect(runner.CheckStatus("Running")).To(BeTrue())
        })

        It("should have a valid IP address", func() {
            // TODO: refactor to function which returns bool
            ip: = runner.RunCommand("ip", true)
            ip = strings.TrimRight(ip, "\n")
            Expect(net.ParseIP(ip)).ToNot(Equal(nil))
        })

        It("should be possible to run ssh", func() {

            expectedStr: = "hello"
            sshCmdOutput: = runner.RunCommand("ssh echo " + expectedStr, true)

            Expect(sshCmdOutput).To(ContainSubstring(expectedStr))
        })

        It("should be possible to set Docker environment variables", func() {

            profileContents: = runner.RunCommand("ssh cat /var/lib/boot2docker/profile", true)

            Expect(profileContents).To(ContainSubstring("FOO=BAR"))
            Expect(profileContents).To(ContainSubstring("BAZ=BAT"))
        })
    })
})
```

**Comment by hferentschik** [comment]
For me the syntax is really mixing test description and impl details a lot.

**Comment by gbraad** [comment]
@hferentschik mixing test description and implementation does not have to be an issue, and in the case of ginkgo you just need to be aware that some tests should better be abstracted in general util functions, as it improves to the readability... but we are not creating an API per se, but rather a CLI tool. for this it feels the gherkin approach with `godog` feels as a better approach, as the focus can be on the commands and the expected output from the test description.

**Comment by hferentschik** [comment]
Looking at the two setups, I still feel that the Cucumber approach has a better separation of concerns and a higher/better abstraction level. 

So let's go ahead with go-dog. The plan would be as follow:

* Create some basic tests outlining the direction 
* Provide some common step definitions which can be re-used moving forward
* Delete existing tests

Once we have the basic harness in place, it will serve as starting point for extending the tests as we go now.

**Comment by budhrg** [comment]
> So let's go ahead with go-dog.

:+1:

**Comment by gbraad** [comment]
@budhrg Circle-CI complains about API rate limit, and I have an issue with adding the dependency to `glide.yaml`; very slow to test. mostly related to the checkout of Kubernetes #168. Any suggestions ?

**Comment by gbraad** [comment]
On CentOS-CI we set the driver as follows
```
05:07:12 + VM_DRIVER=kvm
```

but in the test we explicitly mention:
```
05:08:34: Cmd:start --vm-driver=virtualbox --docker-env=FOO=BAR --docker-env=BAZ=BAT
```

This means the `minishift` command will download the oc binary and b2d iso, but will fail the test. What do we want to do? (`flags.go` does not override this, as it is mentioned in the feature)

**Comment by agajdosi** [comment]
@gbraad I got some ideas looking onto your .feature file. I think it is really needed to specify minishift state in when statement so it is clear that scenario covers start from non existing state or from stopped state or it covers behaviour when starting on running instance of minishift. Next two ideas are more about our attitude, are moc speculative: features could be written from application perspective, for example instead of `I need to be able to bring up a test environment` it could be written as `Minishift is able to start VM successfully.` Last one is that instead of direct commands in scenarios there could be more readable descriptions of behaviour. And concrete commands would be then covered in _test.go code. WDYT?

I am including code from my yesterdays short practice of gherkin: 
```
Feature: Minishift can start virtual machine
  Minishift starts VM successfully on every provisioner (vm-driver) supported for given platform.
  Iso image for creation of VM will be downloaded automatically or if custom path for iso image
  is specified, then minishift uses this image.
  Minishift is able to start VM with every iso image which is oficially supported by minishift.   
  Properties of VM can be further altered within start command by several attributes, key of them are:
  number of virtual CPUs, size of operation memory and disk size allocated for VM.
 
  Background: minishift instance does not exist
    Given status of minishift is does not exist

  Scenario: Minishift starts with native values
    Minishift is able to start VM when any flag is given by user,
    meaning minishift will use native values for platform on which it runs.
    When user starts minishift
    And no flag is specified
    Then status of minishift is running
    And user is able to ssh into the VM

  Scenario Outline: Minishift starts with iso image specified
    Minishift is able to start VM when supported iso image is specified.
    Source for this image can be given as url or as local file.
    When user starts minishift
    And iso-url flag is specified for <distribution> which source is <source_type>
    Then status of minishift is running
    And user is able to ssh into VM

    Examples:
      | distribution | path  |
      | boot2docker  | local |
      | boot2docker  | url   |
      | CentOS       | local |
      | CentOS       | url |

  Scenario Outline: Minishift starts with non native properties for VM
    Minishift is able to start VM when flags for size of operation memory,
    number of CPUs and disk size are given.
    Minishift can start with all combinations of valid values for these flags.
    When user starts minishift
    And memory flag is <memory>, cpus flag is <cpus> and disk-size is <disk-size>
    Then status of minishift is running
    And user is able to ssh into VM

    Examples:
      | memory | cpus   | disk-size |
      | native | native | 30g       |
      | native | 3      | native    |
      | native | 3      | 30g       |
      | 4096   | native | native    |
      | 4096   | native | 30g       |
      | 4096   | 3      | native    |
      | 4096   | 3      | 30g       |
```

**Comment by gbraad** [comment]
@agajdosi a lot of good suggestions. We will incorporate some of these in later increments. At the moment I will implement some of the changes that were suggested on the commit.

**Comment by hferentschik** [comment]
Merged via pull request #433
